### PR TITLE
[#160248419, #160248295, #160248187] Build out API endpoints for questions and answers

### PR DESCRIPTION
**Pull request by jillo-abdullahi**
**What does this PR do?**
Add tests and endpoints for questions and answers

**Description of tasks to be completed?**
`GET /questions`
`GET /questions/<questionId>`
`POST /questions`
`POST /questions/<questionId>/answers`

**How should these be manually tested?**
After cloning the repo, cd into it and run `nosetests`
Using postman test all the endpoints above.

**What are the relevant pivotal tracker stories?**
#160248419, #160248295, #160248187, #160248320

### CODE REVIEW

File: app/v1/views/auth.py [hunk 1]
```diff
@@ -28,8 +27,6 @@ def user_signup():
         response = jsonify({"Error": "Fields should not be empty"})
```
Comment (pndemo) [inline app/v1/views/auth.py]: Each field should be validated with the most appropriate error message(s) being returned for that(those) field(s). The validation functions should be abstracted from the views.

File: app/v1/views/auth.py [hunk 2]
```diff
@@ -38,6 +35,7 @@ def user_signup():
                 {"Error": "A user with username, {} already exists".format(user_info['username'])})
             return response, 400
 
+    user_info["password"] = generate_password_hash(user_info["password"])
     user_instance.save(user_info)
     response = jsonify({"Users": existing_users,
```
Comment (pndemo) [inline app/v1/views/auth.py]: Please ensure each response has a status code.

File: app/v1/views/questions.py [hunk 3]
```diff
@@ -2,48 +2,63 @@
...
 def post_question():
     """method to post a new question"""
     question_info = request.get_json()
 
-    return jsonify({"Question": question_info})
-
-
-@questions.route('/', methods=['GET'])
+    # Check that all fields have been provided
+    if check_keys(question_info, 2):
```
Comment (pndemo) [inline app/v1/views/questions.py]: Change this implementation to be able to tell the user which field(s) has(have) not been provided.

File: app/v1/views/questions.py [hunk 4]
```diff
@@ -2,48 +2,63 @@
...
+        response = jsonify(
+            {"Error": "Please provide question title and description"})
+        return response, 400
+
+    # Check for empty values
+    if check_empty_dict(question_info):
```
Comment (pndemo) [inline app/v1/views/questions.py]: Change this implementation to be able to tell the user which field is empty or invalid. Define and use validation functions as suggested above.

File: app/v1/views/questions.py [hunk 5]
```diff
@@ -2,48 +2,63 @@
...
+        response = jsonify({"Error": "Fields must not be empty"})
+        return response, 400
+
+    # Check if question's already been asked
```
Comment (pndemo) [inline app/v1/views/questions.py]: Abstract this to a question title validation function.

File: app/v1/views/answers.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+answer_instance = Answer()
+question_instance = Question()
+
+
+@answers.route('/<question_id>/answers', methods=['POST'])
+def post_answer(question_id):
+    """method to post a new answer"""
+    answer_info = request.get_json()
+
+    # Check that all fields have been provided
```
Comment (pndemo) [inline app/v1/views/answers.py]: Checking if an answer has been provided and checking if empty should be abstracted into one validation function.

File: tests/test_questions.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    """Class to test for questions"""
+
+    def setUp(self):
+        """Method to create app and set up test client"""
+        self.app = create_app("testing")
+        self.app = self.app.test_client()
+
+    def test_user_can_post_question(self):
+        """Method to test if a new question can be added"""
+        response = self.add_question()
```
Comment (pndemo) [inline tests/test_questions.py]: Please note that only test case methods can be defined in your test class. You can use class variables that should be defined in the `setUp` method.

File: tests/test_questions.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        self.assertEqual(response.status_code, 201)
```
Comment (pndemo) [inline tests/test_questions.py]: Assertions should be done for response messages also.

Comment (pndemo) [reply tests/test_questions.py]: Please note that this applies for other test cases that have been defined in a similar manner.

File: tests/test_answers.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
+"""All tests for posting answers to questions"""
+
+import unittest
+import json
+
+from app import create_app
+
+
+class TestQuestions(unittest.TestCase):
```
Comment (pndemo) [inline tests/test_answers.py]: This class should be renamed something representative of what it does. I.e. `AnswersTests`.

File: app/v1/views/questions.py (hunk 3 above)
Comment (jillo-abdullahi) [reply app/v1/views/questions.py]: This has been implemented.

File: app/v1/views/questions.py (hunk 4 above)
Comment (jillo-abdullahi) [reply app/v1/views/questions.py]: This has been implemented.

File: app/v1/views/questions.py (hunk 5 above)
Comment (jillo-abdullahi) [reply app/v1/views/questions.py]: Done.

File: app/v1/views/answers.py (hunk 6 above)
Comment (jillo-abdullahi) [reply app/v1/views/answers.py]: This has been done.

File: tests/test_questions.py (hunk 7 above)
Comment (jillo-abdullahi) [reply tests/test_questions.py]: Done. Deleted non-test method from class and created class variable.

File: tests/test_questions.py (hunk 8 above)
Comment (jillo-abdullahi) [reply tests/test_questions.py]: Done. Tested response messages for all post requests.

File: tests/test_answers.py (hunk 9 above)
Comment (jillo-abdullahi) [reply tests/test_answers.py]: My bad! Class name has been updated.

**Pull request merged**